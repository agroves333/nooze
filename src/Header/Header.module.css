.container {
    margin-top: 75px;
}

.nav {
    box-shadow: 0 1px 5px 0 rgba(0,0,0,0.5);
}

.logo {
    font-family: 'Comfortaa', cursive;
    font-size: 26px;
}

.addFavorite, .dropdownItem {
    cursor: pointer;
}

.searchButton {
    outline: none;
    box-shadow:none !important;
    border:1px solid #ccc !important;
}

.history, .savedSearches {
    height: 300px;
    overflow-y: scroll;
}
