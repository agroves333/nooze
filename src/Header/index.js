import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Navbar,
  NavbarBrand,
  Nav,
  NavItem,
  UncontrolledDropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem,
  Collapse,
  Input,
  InputGroup,
  InputGroupAddon,
  Button,
  UncontrolledTooltip,
} from 'reactstrap';

import Octicon, { History } from '@githubprimer/octicons-react'

import styles from './Header.module.css';
import moment from "moment/moment";

class Header extends Component {
  
  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
    };
    this.searchInput = React.createRef();
  }
  
  /**
   * Perform search when Enter key is pressed
   * @param e Keyup Event
   */
  handleSearchKeyUp = (e) => {
    e.preventDefault();
    if (e.keyCode === 13) {
      this.props.onSearch(this.props.query);
    }
  };

  toggle = () => {
    this.setState({
      isOpen: !this.state.isOpen
    });
  };
  
  /**
   * Save search to local storage and state.
   */
  saveSearch = () => {
    let savedSearches = localStorage.getItem('savedSearches');
    savedSearches = savedSearches ? JSON.parse(savedSearches) : [];
    savedSearches.push(this.props.query);
    this.setState({
      savedSearches,
    });
    localStorage.setItem('savedSearches', JSON.stringify(savedSearches));
  };
  
  /**
   * Render Saved Searches dropdown
   */
  renderSavedSearches() {
    let savedSearches = localStorage.getItem('savedSearches');
    savedSearches = savedSearches ? JSON.parse(savedSearches).reverse() : [];
    return savedSearches.length ? savedSearches.map((savedSearch, key) => {
      return (
        <DropdownItem
          key={`saved-search-${key}`}
          onClick={() => this.props.onSearch(savedSearch)}
          className={styles.dropdownItem}
        >
          {savedSearch}
        </DropdownItem>
      )
    }) : (
        <DropdownItem
          key={`saved-search-empty`}
          className={styles.dropdownItem}
        >
          No saved searches
        </DropdownItem>
    )
  }
  
  renderHistory() {
    return this.props.searchHistory &&
      this.props.searchHistory
        .reverse()
        .map((item, key) => {
          return (
            <DropdownItem
              key={`favorite-${key}`}
              onClick={() => this.props.onSearch(item.query)}
              className={styles.dropdownItem}
            >
              <div>
                <span className="mr-2">"{item.query}"</span>
                <small className="text-muted">{moment(item.time).fromNow()}</small>
              </div>
            </DropdownItem>
          )
        });
  }

  render() {
    return (
      <div>
        <Navbar className={styles.nav} color="light" light expand="xs" fixed="top">
          <NavbarBrand
            className={styles.logo}
            href="/">
            <span className="d-none d-sm-inline">[nooze]</span>
            <span className="d-inline d-sm-none">[n]</span>
          </NavbarBrand>
          <Collapse isOpen={this.state.isOpen} navbar>
            <Nav className="ml-auto" navbar>
              <NavItem className="d-flex mr-4">
                <InputGroup className="align-items-center">
                  <InputGroupAddon
                    className={styles.addFavorite}
                    addonType="prepend"
                  >
                    <Button
                      id="saveSearchButton"
                      color="secondary"
                      className={styles.saveSearchButton}
                      onClick={this.saveSearch}
                    >
                      ★
                    </Button>
                    <UncontrolledTooltip placement="bottom" target="saveSearchButton">
                      Save search
                    </UncontrolledTooltip>
                  </InputGroupAddon>
                  <Input
                    ref={this.searchInput}
                    placeholder="Search nooze"
                    value={this.props.query}
                    onChange={this.props.onUpdateQuery}
                    onKeyUp={this.handleSearchKeyUp}
                  />
                  <InputGroupAddon addonType="append">
                    <Button
                      color="primary"
                      type="submit"
                      className={styles.searchButton}
                      onClick={() => this.props.onSearch(this.props.query)}
                    >
                      Search
                    </Button>
                  </InputGroupAddon>
                </InputGroup>
              </NavItem>
              
              <UncontrolledDropdown nav inNavbar>
                <DropdownToggle nav>
                  ★ <span className="d-none d-md-inline">Saved Searches</span>
                </DropdownToggle>
                <DropdownMenu right className={styles.savedSearches}>
                  {this.renderSavedSearches()}
                </DropdownMenu>
              </UncontrolledDropdown>
              
              <UncontrolledDropdown nav inNavbar>
                <DropdownToggle nav>
                  <Octicon icon={History} /> <span className="d-none d-md-inline">History</span>
                </DropdownToggle>
                <DropdownMenu right className={styles.history}>
                  {this.renderHistory()}
                </DropdownMenu>
              </UncontrolledDropdown>
            </Nav>
          </Collapse>
        </Navbar>
      </div>
    );
  }
}

Header.propTypes = {
  query: PropTypes.string,
  onSearch: PropTypes.func,
  onUpdateQuery: PropTypes.func,
  searchHistory: PropTypes.array,
};

Header.defaultProps = {
  onSearch: () => {},
  onUpdateQuery: () => {},
  searchHistory: [],
};

export default Header;
