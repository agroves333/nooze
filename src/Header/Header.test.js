import React from 'react';
import { render, shallow } from 'enzyme';
import Header from './index';

const savedSearches = ['fake-search', 'fake-search', 'fake-search'];

const history = [
  {
    query: 'fake-query',
    time: 1234567890,
  },
  {
    query: 'fake-query',
    time: 1234567890,
  },
  {
    query: 'fake-query',
    time: 1234567890,
  },
];
describe('Header', () => {
  
  beforeEach(() => {
    const localStorageMock = (() => {
      let store = {};
    
      return {
        getItem: key => {
          return store[key] || null;
        },
        setItem: (key, value) => {
          store[key] = value.toString();
        },
        clear: () => {
          store = {};
        }
      };
    })();
  
    Object.defineProperty(window, 'localStorage', {
      value: localStorageMock
    });
  });
  
  it('renders Header', () => {
    const wrapper = shallow(<Header/>);
    expect(wrapper).toMatchSnapshot();
  });
  
  it('renders history dropdown with 3 items', () => {
    const wrapper = render(<Header searchHistory={history}/>);
    const dropdown = wrapper.find('li.dropdown').eq(1);
    const dropdownItems = dropdown.find('button.dropdown-item');
    expect(dropdownItems).toHaveLength(3);
  });
  
  it('renders saved searches dropdown with 3 items', () => {
    localStorage.setItem('savedSearches', JSON.stringify(savedSearches));
    const wrapper = render(<Header/>);
    const dropdown = wrapper.find('li.dropdown').eq(0);
    const dropdownItems = dropdown.find('button.dropdown-item');
    expect(dropdownItems).toHaveLength(3);
  });
});
