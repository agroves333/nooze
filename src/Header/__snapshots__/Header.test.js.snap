// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders Header 1`] = `
<div>
  <Navbar
    className="nav"
    color="light"
    expand="md"
    fixed="top"
    light={true}
    tag="nav"
  >
    <NavbarBrand
      className="logo"
      href="/"
      tag="a"
    >
      nooze
    </NavbarBrand>
    <Collapse
      appear={false}
      enter={true}
      exit={true}
      in={false}
      isOpen={false}
      mountOnEnter={false}
      navbar={true}
      onEnter={[Function]}
      onEntered={[Function]}
      onEntering={[Function]}
      onExit={[Function]}
      onExited={[Function]}
      onExiting={[Function]}
      tag="div"
      timeout={350}
      unmountOnExit={false}
    >
      <Nav
        className="ml-auto"
        navbar={true}
        tag="ul"
        vertical={false}
      >
        <NavItem
          className="d-flex mr-4"
          tag="li"
        >
          <InputGroup
            className="align-items-center"
            tag="div"
          >
            <InputGroupAddon
              addonType="prepend"
              className="addFavorite"
              onClick={[Function]}
              tag="div"
            >
              <Button
                className="saveSearchButton"
                color="secondary"
                id="saveSearchButton"
                tag="button"
              >
                ★
              </Button>
              <UncontrolledTooltip
                placement="bottom"
                target="saveSearchButton"
              >
                Save search
              </UncontrolledTooltip>
            </InputGroupAddon>
            <Input
              onChange={[Function]}
              onKeyUp={[Function]}
              placeholder="Search nooze"
              type="text"
              value="Elon Musk"
            />
            <InputGroupAddon
              addonType="append"
              onClick={[Function]}
              tag="div"
            >
              <Button
                className="searchButton"
                color="primary"
                onClick={[Function]}
                tag="button"
                type="submit"
              >
                Search
              </Button>
            </InputGroupAddon>
          </InputGroup>
        </NavItem>
        <UncontrolledDropdown
          inNavbar={true}
          nav={true}
        >
          <DropdownToggle
            aria-haspopup={true}
            color="secondary"
            nav={true}
          >
            ★ Saved Searches
          </DropdownToggle>
          <DropdownMenu
            className="savedSearches"
            flip={true}
            right={true}
            tag="div"
          >
            <DropdownItem
              className="dropdownItem"
              key="saved-search-empty"
              tag="button"
              toggle={true}
            >
              No saved searches
            </DropdownItem>
          </DropdownMenu>
        </UncontrolledDropdown>
        <UncontrolledDropdown
          inNavbar={true}
          nav={true}
        >
          <DropdownToggle
            aria-haspopup={true}
            color="secondary"
            nav={true}
          >
            <Octicon
              className="octicon"
              icon={[Function]}
              size={16}
              verticalAlign="text-bottom"
            />
             History
          </DropdownToggle>
          <DropdownMenu
            className="history"
            flip={true}
            right={true}
            tag="div"
          />
        </UncontrolledDropdown>
      </Nav>
    </Collapse>
  </Navbar>
</div>
`;
